digest/short.go

```diff
@@ -0,0 +1,186 @@
+
+// NewShortTable creates an empty table of digests
+// which may have digests added.
+func NewShortTable() *ShortTable {
+	return &ShortTable{
+		entries: digestEntries{},
+	}
+}
+
+func checkShortMatch(alg, hex, shortAlg, shortHex string) bool {
+	if len(hex) == len(shortHex) {
+		if hex != shortHex {
+			return false
+		}
+		if len(shortAlg) > 0 && alg != shortAlg {
+			return false
+		}
+	} else if !strings.HasPrefix(hex, shortHex) {
+		return false
+	} else if len(shortAlg) > 0 && alg != shortAlg {
+		return false
+	}
+	return true
+}
+
+// Lookup looks for a digest matching the given short code. If
+// no digests could be found ErrShortCodeNotFound will be returned
+// with an empty digest value. If multiple matches are found
+// ErrShortCodeAmbiguous will be returned with an empty digest
+// value.
+func (dst *ShortTable) Lookup(short string) (Digest, error) {
+	if len(dst.entries) == 0 {
+		return "", ErrShortCodeNotFound
+	}
+	var (
+		searchFunc func(int) bool
+		alg        string
+		hex        string
+	)
+	dgst, err := ParseDigest(short)
+	if err == ErrDigestInvalidFormat {
+		hex = short
+		searchFunc = func(i int) bool {
+			return dst.entries[i].val >= short
+		}
+	} else {
+		hex = dgst.Hex()
+		alg = dgst.Algorithm()
+		searchFunc = func(i int) bool {
+			if dst.entries[i].val == hex {
+				return dst.entries[i].alg >= alg
+			}
+			return dst.entries[i].val >= hex
+		}
+	}
+	idx := sort.Search(len(dst.entries), searchFunc)
+	if idx == len(dst.entries) || !checkShortMatch(dst.entries[idx].alg, dst.entries[idx].val, alg, hex) {
+		return "", ErrShortCodeNotFound
+	}
+	if dst.entries[idx].alg == alg && dst.entries[idx].val == hex {
+		return dst.entries[idx].digest, nil
+	}
+	if idx+1 < len(dst.entries) && checkShortMatch(dst.entries[idx+1].alg, dst.entries[idx+1].val, alg, hex) {
+		return "", ErrShortCodeAmbiguous
+	}
+
+	return dst.entries[idx].digest, nil
+}
+
+// Add adds the given digests to the table. An error will be returned
+// if the given digest is invalid. If the digest already exists in the
+// table, this operation will be a no-op.
+func (dst *ShortTable) Add(d Digest) error {
+	if err := d.Validate(); err != nil {
+		return err
+	}
+	entries := append(dst.entries, &digestEntry{alg: d.Algorithm(), val: d.Hex(), digest: d})
+	for i := len(entries) - 1; i > 0; i-- {
+		if !entries.Less(i, i-1) {
+			if entries[i].digest == entries[i-1].digest {
+				for j := i; j < len(entries)-1; j++ {
+					entries.Swap(j, j+1)
+				}
+				entries = entries[:len(entries)-1]
+			}
+			break
+		}
+		entries.Swap(i, i-1)
+	}
+	dst.entries = entries
+	return nil
+}
+
+// DumpTable returns a map of Digest to unique short codes. The
+// length represents the minimum value, the maximum length may be the
+// entire value of digest if uniqueness cannot be achieved without the
+// full value. This function will attempt to make short codes as short
+// as possible to be unique.
+func (dst *ShortTable) DumpTable(length int) map[Digest]string {
```

Comment:
A reasonable length is contextual, this function only provides options. Could add a strict length mode which returns an error if an ambiguity is detected. This implementation is smart enough to always return unique values no matter the length given.